### PR TITLE
Add PerplexityMetric for language model evaluation

### PR DESCRIPTION
**Pull request by TheDarkchip**
### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Addresses the need for perplexity evaluation in language model training, commonly requested for LLM development workflows.

### Changes

**Summary**: Adds a comprehensive PerplexityMetric for evaluating language models in the Burn framework, with mathematically correct batch aggregation and seamless integration.

**Core Implementation:**
- **`PerplexityMetric<B: Backend>`**: Main metric implementing `Metric` and `Numeric` traits
- **`PerplexityInput<B: Backend>`**: Input type for logits and target tokens  
- **`PerplexityState`**: Custom state for mathematically correct cross-batch aggregation
- **Integration**: Added to `examples/text-generation` training loop

**Key Mathematical Fix:**
Unlike other metrics that can be averaged, perplexity requires special aggregation:
- Accumulates total negative log-likelihood across all tokens
- Accumulates total number of effective tokens (excluding padding)
- Computes final perplexity as `exp(total_nll / total_tokens)` only at the end

This ensures epoch-level perplexity accuracy regardless of batch sizes or padding variations.

**Features:**
- ✅ **Padding Support**: Configurable pad token handling excluding padding tokens
- ✅ **Correct Aggregation**: Custom `PerplexityState` prevents mathematical errors from averaging perplexities
- ✅ **API Consistency**: Follows Burn's metric patterns (`AccuracyMetric`, `LossMetric`)
- ✅ **Backend Agnostic**: Works with any Burn backend
- ✅ **Integration Ready**: Works with existing `ClassificationOutput` via `Adaptor` trait

**Files Modified:**
- `crates/burn-train/src/metric/perplexity.rs` (new) - Core implementation
- `crates/burn-train/src/metric/mod.rs` - Export new metric
- `crates/burn-train/src/metric/state.rs` - Add getter methods for `FormatOptions`
- `crates/burn-train/src/learner/classification.rs` - Add `PerplexityInput` adaptor
- `examples/text-generation/src/training.rs` - Integration example

### Testing

**Comprehensive test coverage (5 unit tests):**

```rust
cargo test -p burn-train perplexity
running 5 tests
test metric::perplexity::tests::test_perplexity_perfect_prediction ... ok
test metric::perplexity::tests::test_perplexity_uniform_prediction ... ok  
test metric::perplexity::tests::test_perplexity_wrong_prediction ... ok
test metric::perplexity::tests::test_perplexity_with_padding ... ok
test metric::perplexity::tests::test_perplexity_multi_batch_aggregation ... ok

test result: ok. 5 passed; 0 failed; 0 ignored
```

**Test scenarios covered:**
1. **Perfect predictions** - Low perplexity (~1.0)
2. **Uniform predictions** - Expected perplexity (~vocab_size)
3. **Wrong predictions** - High perplexity (>10)
4. **Padding handling** - Proper masking of pad tokens
5. **Multi-batch aggregation** - Verifies mathematical correctness vs single-batch equivalent

**Quality assurance:**
- Code formatting: ✅ `cargo fmt --check`
- Linting: ✅ `cargo clippy --no-deps -- --deny warnings`  
- Documentation: ✅ Comprehensive docstrings with mathematical explanations
- No breaking changes: ✅ Purely additive functionality

**Usage example:**
```rust
.metric_train_numeric(PerplexityMetric::new().with_pad_token(tokenizer.pad_token()))
.metric_valid_numeric(PerplexityMetric::new().with_pad_token(tokenizer.pad_token()))
```

The metric will be displayed alongside existing accuracy and loss metrics during training.

### CODE REVIEW

File: crates/burn-train/src/metric/perplexity.rs [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+/// - Compute perplexity as exp(total_nll / total_tokens) only at the end
+#[derive(Clone)]
+struct PerplexityState {
+    /// Sum of negative log-likelihood across all tokens
+    sum_nll: f64,
+    /// Total number of effective tokens (excluding padding)
+    total_tokens: usize,
+    /// Current batch perplexity (for display purposes)
+    current: f64,
+}
```
Comment (laggui) [inline crates/burn-train/src/metric/perplexity.rs:23]: This will be helpful for other metrics that require such a running state. We could probably generalize it in the future 🙂

**Review by laggui — APPROVED** [review]
Thanks for adding this metric! Implementation looks good to me
